core/src/sumeragi/network_topology.rs

```diff
@@ -5,341 +5,182 @@
-            .map(|peer| peer.public_key)
-            .collect();
+        let mut public_keys = Vec::new();
+        for role in roles {
+            let role_public_keys = if self.max_faults() == 0 {
+                match role {
+                    Role::Leader => vec![self.sorted_peers[0].public_key.clone()],
+                    Role::ValidatingPeer => {
+                        if self.sorted_peers.len() > 2 {
+                            vec![self.sorted_peers[1].public_key.clone()]
+                        } else {
+                            vec![]
+                        }
+                    }
+                    Role::ProxyTail => {
+                        if self.sorted_peers.len() == 2 {
+                            vec![self.sorted_peers[1].public_key.clone()]
+                        } else if self.sorted_peers.len() > 2 {
+                            vec![self.sorted_peers[2].public_key.clone()]
+                        } else {
+                            vec![]
+                        }
+                    }
+                    Role::ObservingPeer => {
+                        if self.sorted_peers.len() == 4 {
+                            vec![self.sorted_peers[3].public_key.clone()]
+                        } else {
+                            vec![]
+                        }
+                    }
+                }
+            } else {
+                match role {
+                    Role::Leader => vec![self.sorted_peers[0].public_key.clone()],
+                    Role::ValidatingPeer => self.sorted_peers[1..(self.min_votes_for_commit() - 1)]
+                        .iter()
+                        .map(|peer_id| peer_id.public_key.clone())
+                        .collect(),
+                    Role::ProxyTail => vec![self.sorted_peers[self.min_votes_for_commit() - 1]
+                        .public_key
+                        .clone()],
+                    Role::ObservingPeer => self.sorted_peers[self.min_votes_for_commit()..]
+                        .iter()
+                        .map(|peer_id| peer_id.public_key.clone())
+                        .collect(),
+                }
+            };
+            public_keys.extend(role_public_keys);
+        }
         signatures
             .into_iter()
             .filter(|signature| public_keys.contains(signature.public_key()))
             .cloned()
             .collect()
     }
-
-    /// Sorted peers that this topology has.
-    pub fn sorted_peers(&self) -> &[PeerId] {
-        &self.sorted_peers
+    /// What role does this peer have in the topology. If it is not in the toplogy
+    /// the function will return `Role::ObservingPeer`.
+    // This lint is a bad suggestion.
+    #[allow(clippy::option_if_let_else)]
+    pub fn role(&self, peer_id: &PeerId) -> Role {
+        if let Some(index) = self
+            .sorted_peers
+            .iter()
+            .position(|p| p.public_key == peer_id.public_key)
```

Comment:
```suggestion
            .position(|p| p == peer_id)
```

peers are compared by public keys

Reply:
Good to know thanks.